NAS2D/Renderer/Window.h

```diff
@@ -85,6 +85,18 @@ class Window {
 
 	virtual void SetDimensionsAndPosition(std::pair<int, int> newClientDimensions, std::pair<int, int> newPosition) noexcept = 0;
 
+	/**
+	 * @brief Set the native OS-specific window icon resource.
+	 * @param iconResource: Handle to the icon resource.
+	 * @return 
+	*/
+	virtual void Icon(void* iconResource) noexcept = 0;
+
+	/**
+	 * @brief Get the native OS-specific window icon resource.
+	 * @return void* handle to the native OS-specific window icon resource.
+	*/
+	[[nodiscard]] virtual void* Icon() const noexcept = 0;
```

Comment:
We might want to use `Image` references, rather than `void*`. At least that would make sense for set. Implementing get might be more awkward.

Reply:
These don't actually use images. For Windows it can either be a reference to a Resource file Resource ID, a handle inside the actual executable, an OEM Icon, or they can be a LPCSTR string path to the file if loading from disk or a string of the form `#xxx` indicating a resource ID.

See: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadimagea

From what I understand on Linux, it's a reference inside of, or to, a `.desktop` file.

The `void*` is for avoiding referencing the OS-specific types in the base class and the fact that, at least for Windows, the argument can be one of a couple different data formats. The intended derived types `WindowWindows` and `WindowLinux` and `WindowMacOS` should handle the appropriate types inside the function:

This also exemplifies the use of `nativeHandle()` through the interface.

I would figure Linux to be similar.

Example of loading off disk:
```cpp
WindowWindows::icon(void* iconResource)
{
    HWND hWnd = static_cast<HWND>(nativeHandle());
    bool isResourceIdOrPath = !!IS_INTRESOURCE(iconResource);
    if(isResourceIdOrPath)
    { //Resource ID or OEM Resource ID
        //iconResource would be a custom resource ID from a .rc Resource file or one of the following defined in WinUser.h:
        //OIC_SAMPLE, OIC_HAND, OIC_QUES, OIC_BANG, OIC_NOTE, OIC_WINLOGO, OIC_WARNING, OIC_ERROR, OIC_HAND, OIC_INFORMATION, OIC_SHIELD
        auto resourceAsInt = static_cast<int>(iconResource);
        //0,0 and fuLoad != LR_DEFAULTSIZE  says use resource size
        HICON hIcon = ::LoadImageA(nullptr, MAKEINTRESOURCE(resourceAsInt), IMAGE_ICON, 0, 0, 0);
        ::SendMessage(hWnd, WM_SETICON, ICON_SMALL, (LPARAM)hIcon);
    }
    else
    { //Path on disk to file or resource ID string
        const char* imagePath = reinterpret_cast<const char*>(iconResource);
        if(imagePath[0] == '#')
        {
            std::string imagePathAsResourceId{imagePath};
            auto resourceIdAsInt = static_cast<int>(std::stoi(imagePathAsResourceId.substr(1)));
            //0,0 and fuLoad != LR_DEFAULTSIZE  implies using the actual resource size. For multi-icon .ico files, the first one is picked if no size is supplied.
            HICON hIcon = ::LoadImageA(nullptr, MAKEINTRESOURCE(resourceAsInt), IMAGE_ICON, 0, 0, LR_LOADFROMFILE);
            ::SendMessage(hWnd, WM_SETICON, ICON_SMALL, (LPARAM)hIcon);
        }
        else
        {
            //0,0 and fuLoad != LR_DEFAULTSIZE  implies using the actual resource size. For multi-icon .ico files, the first one is picked if no size is supplied.
            HICON hIcon = ::LoadImageA(nullptr, imagePath, IMAGE_ICON, 0, 0, LR_LOADFROMFILE);
            ::SendMessage(hWnd, WM_SETICON, ICON_SMALL, (LPARAM)hIcon);
        }
    }
}
```

Reply:
Note to self: The documentation for WM_SETICON is awful.

Reply:
That is one of the challenges here. Each platform has a rather different way of doing this, and there isn't really a way to have a single function signature that can handle all platforms, and still be type safe without need of any casts. That should maybe give pause about putting such a method in a base class interface.

I'm afraid I don't have a good suggestion for this.

----

I should probably mention that I would really like to stay away from platform specific code. It's very hard for a single developer to properly test and update code if it runs differently on their local development platform from other platforms. I would have quite some difficulty in maintaining Windows specific code. I would rather see subtypes specific to generic windowing libraries, such as SDL, that abstract away platform differences. There may be other windowing libraries that could be used instead of SDL. I suppose someone could write a Windows API specific one. Though I would like to avoid putting that in the NAS2D library. Honestly, I only expect us to support SDL. Anything else would be a client concern. I don't want to take support any further than having a generic enough API that developers could derive their own classes to handle other libraries.

----

It might be possible to kind of abstract away differences with a generic `Image` type. I don't think it's a perfect abstraction though. The current `RendererOpenGL::window_icon` method has SDL specific code in it, which is similar to SDL specific code in the `Image` class. I don't have any great ideas for how a supposed generic `Image` class could be used by different windowing libraries though. I imagine it would be tied to a specific library, such as `SdlImage`. At that point, you could use `friend` declarations between related classes that depend on the same underlying library. Of course that means splitting a lot of the resource classes into a generic base type and a library specific version. I'm not really looking forward to doing that.

Reply:
Just tested this with my own code seen below, loading an `.ico` file at the point of window registration with the OS makes this trivial and it "just works" on Windows. There's no need for a `SetIcon` or `GetIcon`. Changing them at runtime is not a thing anyway.

```cpp
bool Window::Register() noexcept {
    _hInstance = GetModuleHandle(nullptr);
    memset(&_wc, 0, sizeof(_wc));
    auto window_class_name = "Simple Window Class";
    _wc.cbSize = sizeof(_wc);
    _wc.style = CS_OWNDC | CS_HREDRAW | CS_VREDRAW | CS_DBLCLKS;
    _wc.lpfnWndProc = EngineMessageHandlingProcedure;
    _wc.cbClsExtra = 0;
    _wc.cbWndExtra = 0;
    _wc.hInstance = _hInstance;
    //Enforce downstream projects must put their custom icon.ico file in a specific spot to be found by the engine.
    auto path = FileUtils::GetKnownFolderPath(FileUtils::KnownPathID::GameData) / "Resources/Icons/icon.ico";
    //.ico file contains multiple image sizes. Load image Big image that goes in the taskbar and Alt+Tab window-switching list.
    HICON iconTaskbar = (HICON)::LoadImage(nullptr, path.string().c_str(), IMAGE_ICON, 64, 64, LR_LOADFROMFILE);
    _wc.hIcon = iconTaskbar;
    _wc.hCursor = ::LoadCursor(NULL, IDC_ARROW);
    _wc.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
    _wc.lpszMenuName = NULL;
    _wc.lpszClassName = window_class_name;
    //Load the small version that goes in the title bar.
    HICON iconTitlebar = (HICON)::LoadImage(nullptr, path.string().c_str(), IMAGE_ICON, 16, 16, LR_LOADFROMFILE);
    _wc.hIconSm = iconTitlebar;
    return 0 != RegisterClassEx(&_wc);
}
```

The above makes the icon appear in all four places: The title bar, the task bar, the alt+tab list, and the .exe image. There's no need for dealing with `.rc` resource files and no WM_SETICON message is ever sent (and I've never seen a program change them at runtime due to branding familiarity); making setup much easier.

Enforcing specific asset locations to be read by a game engine is not new. Unreal expects assets to be in specific locations in the Content folder to be loaded correctly and it's no different for the project's window icon. Unreal enforces this by specifying the image in the project settings at compile time that is then copied into the correct place and the path is "hard coded" by the code generator for use during window creation.

Reply:
You make a good point about not changing the icon at runtime. It would make reasonable sense to have the icon set once during construction, with no way to change it afterwards.

Reply:
I took a look at the original code from `RendererOpenGL`. It will handle loading icons from the `Filesystem` search path just fine. Seems like a reasonable way to continue setting window icons.

```cpp
void RendererOpenGL::window_icon(const std::string& path)
{
	auto iconData = Utility<Filesystem>::get().read(path);
	SDL_Surface* icon = IMG_Load_RW(SDL_RWFromConstMem(iconData.c_str(), static_cast<int>(iconData.size())), 1);
	if (!icon)
	{
		throw std::runtime_error("Failed to set window icon: " + path + " : " + SDL_GetError());
	}

	SDL_SetWindowIcon(underlyingWindow, icon);
	SDL_FreeSurface(icon);
}
```

I would make sense to add a call to such a method from the constructor, assuming an icon was specified during construction.

I suppose we could leave it as a separate method though, as that enables people to change the icon later, if they want, though would not be required to.

There is currently no way to retrieve the window icon. I'm fine with that. We don't need to support getting the window icon. Presumably they were the ones that set it, so they should already know what it is.

Reply:
I guess this boils down to setting the window icon by filename, rather than by some kind of `Image` reference. I'm fine with such a restriction. It matches existing behaviour, so no loss of functionality.

----

Looking back at the proposed code and my suggestion, I suppose we could generalize to an `Image` reference if we gave the `RendererOpenGL` friend access to the `SDL_Surface` member of an `Image` object. That would give the API more flexibility. Though that's more of a potential future upgrade to the API. I don't think we need to start there. And we might not want to use `friend` access. Probably better to make as minimal of a change as possible when starting out. I think I'd be more comfortable with keeping the icon specified through a filename to start with.

Reply:
> 
> ...The above makes the icon appear in all four places: The title bar, the task bar, the alt+tab list, and the .exe image. There's no need for dealing with `.rc` resource files...

Minor correction to this statement. A resource file and associated resource.h file (generated by visual studio) is required in order to embed the icon into the .exe and to appear as the .exe image. The icon appears in all other places with just an appropriate path provided at window registration time.